addons/website_sale/controllers/main.py

```diff
@@ -1230,3 +1231,69 @@ def products_autocomplete(self, term, options={}, **kwargs):
                 res_product['price'] = FieldMonetary.value_to_html(res_product['price'], monetary_options)
 
         return res
+
+    # --------------------------------------------------------------------------
+    # Products Recently Viewed
+    # --------------------------------------------------------------------------
+    @http.route('/shop/products/recently_viewed', type='json', auth='public', website=True)
+    def products_recently_viewed(self, **kwargs):
+        """
+        Returns list of recently viewed products according to current user and product options
+        """
+        excluded_products = []
+        max_number_of_product_for_carousel = 12
+        current_sale_order = request.website.sale_get_order()
+        if current_sale_order:
+            # Exclude items already in the cart
+            # order_lines = request.env['sale.order.line'].sudo().search([('order_id', '=', current_sale_order.id)])
+            excluded_products += current_sale_order.mapped('line_ids.product_id')
```

Comment:
as explain in pad, it will crash... line_ids does'nt exists, and later (recordset - (list of record)) will crash too.

As well you work with id like previously and browse to have the record set
as well you need to flat the list in recordset